Add local id support

IvanRN opened the pull request:
According to upcoming v1.1 of the JSONAPI the document object
identification is a combination of three values: id, type and
lid (local id). Local id is an id that can be generated when
entity was created but not yet persisted to the DB. In the
JSONAPI it should allow client to match data which he created
and then received back from the server.

IvanRN committed:
Add local id support

According to upcoming v1.1 of the JSONAPI the document object
identification is a combination of three values: id, type and
lid (local id). Local id is an id that can be generated when
entity was created but not yet persisted to the DB. In the
JSONAPI it should allow client to match data which he created
and then received back from the server.

artlo commented on jsonapi/marshal.go:
```diff
@@ -444,17 +456,3 @@ func marshalStruct(data MarshalIdentifier, information ServerInformation) (*Docu
 }
-
-func getStructType(data interface{}) string {
-	entityName, ok := data.(EntityNamer)
```
I think we need to save this functionality for pluralising and using structure name as json API name

IvanRN replied:
restored this functionality

artlo commented on jsonapi/unmarshal.go:
```diff
@@ -10,19 +10,19 @@ import (
 // to-one relations.
 type UnmarshalToOneRelations interface {
-	SetToOneReferenceID(name, ID string) error
+	SetToOneReferenceID(name string, ID *RelationshipData) error
```
why is RelationshipData? if we introduce Identifier struct? It's not handy to have different types for Identifiers

IvanRN replied:
replaced with *Identifier for ToOne and with []Identifier for ToMany

artlo commented on jsonapi/unmarshal.go:
```diff
@@ -227,9 +224,13 @@ func setRelationshipIDs(relationships map[string]Relationship, target UnmarshalI
-	actualType := getStructType(target)
-	if incomingType != actualType {
-		return fmt.Errorf("Type %s in JSON does not match target struct type %s", incomingType, actualType)
+	castedTarget, ok := target.(MarshalIdentifier)
```
if you save method getStructType, you no need to cast here

IvanRN replied:
agree, restored getStructType and removed this casting

artlo commented on api.go:
```diff
@@ -207,23 +207,14 @@ func (api *API) addResource(prototype jsonapi.MarshalIdentifier, source interfac
-		name = resourceType.Name()
 	} else {
 		ptrPrototype = reflect.ValueOf(prototype).Interface()
-		name = resourceType.Elem().Name()
```
save getting name from Go Struct Name if GetID doesn't return name

IvanRN replied:
saved this behaviour (get from struct if Identifier does not have name)

artlo commented on api.go:
```diff
@@ -207,23 +207,14 @@ func (api *API) addResource(prototype jsonapi.MarshalIdentifier, source interfac
 	}
 
-	// check if EntityNamer interface is implemented and use that as name
-	entityName, ok := prototype.(jsonapi.EntityNamer)
```
this also should be saved, it looks similar that is done in function `getStructType `

IvanRN replied:
restored this behaviour

artlo commented on api_interfaces_test.go:
```diff
@@ -120,22 +122,22 @@ var _ = Describe("Test interface api type casting", func() {
 	})
 
 	It("FindAll returns 404 for simple CRUD", func() {
-		req, err := http.NewRequest("GET", "/v1/someDatas", nil)
```
save existing behaviour for api name

artlo commented on examples/model/model_chocolate.go:
```diff
@@ -1,19 +1,28 @@
+}
+
+// GetName to satisfy jsonapi.MarshalIdentifier interface
+func (c Chocolate) GetName() string {
```
This method is not needed anymore

artlo commented on examples/model/model_user.go:
```diff
@@ -18,13 +19,19 @@ type User struct {
+}
+
+// GetName to satisfy jsonapi.MarshalIdentifier interface
+func (u User) GetName() string {
```
This method is not needed anymore

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -14,8 +14,8 @@ type Magic struct {
-func (m Magic) GetID() string {
-	return m.ID.String()
+func (m Magic) GetID() Identifier {
+	return Identifier{ID: m.ID.String(), Name: "magics"}
```
Save possibility ignoring Name in Indentifier

artlo replied:
Not fixed

artlo commented on jsonapi/marshal_composition_test.go:
```diff
@@ -25,7 +29,7 @@ var _ = Describe("Embedded struct types", func() {
 			Expect(err).To(BeNil())
 			Expect(i).To(MatchJSON(`{
 				"data": {
-					"type":"taggedPosts",
```
keep previous behaviour

artlo commented on jsonapi/marshal_composition_test.go:
```diff
@@ -43,7 +47,7 @@ var _ = Describe("Embedded struct types", func() {
 	Context("When unmarshaling objects with struct composition", func() {
 		postJSON := `{
 			"data": {
-				"type": "taggedPosts",
```
the same

IvanRN replied:
restored

artlo commented on jsonapi/marshal_different_to_many_test.go:
```diff
@@ -71,7 +72,7 @@ var _ = Describe("Marshalling toMany relations with the same name and different
             			]
           			}
         		},
-        		"type": "manyParents"
```
the same

IvanRN replied:
restored

artlo commented on jsonapi/marshal_enum_test.go:
```diff
@@ -54,24 +54,25 @@ func (s *PublishStatus) UnmarshalJSON(data []byte) error {
 var _ = Describe("Custom enum types", func() {
 	status := StatusPublished
 	statusValue := "published"
-	singleJSON := []byte(`{"data":{"id": "1", "type": "enumPosts", "attributes": {"title":"First Post","status":"published"}}}`)
```
keep previous behaviour for API name

IvanRN replied:
restored

artlo commented on jsonapi/marshal_test.go:
```diff
@@ -58,7 +58,7 @@ var _ = Describe("Marshalling", func() {
 			Expect(err).To(BeNil())
 			Expect(i).To(MatchJSON(`{
 				"data": {
-					"type": "simplePosts",
```
keep previous behaviour

IvanRN replied:
restored

artlo commented on jsonapi/marshal_test.go:
```diff
@@ -98,7 +98,7 @@ var _ = Describe("Marshalling", func() {
 			Expect(i).To(MatchJSON(`{
 				"data": [
 					{
-						"type": "simplePosts",
```
the same

IvanRN replied:
restored

IvanRN committed:
Add name in Identifier

Identifier is a usefull object for which has id, local id and entity
name/type. This update should fit the upcoming jsonapi v1.1 release
where each object identifier hat this three attributes.

artlo commented on api_test.go:
```diff
@@ -37,25 +37,27 @@ func (m *requestURLResolver) SetRequest(r http.Request) {
-func (i invalid) GetID() string {
-	return "invalid"
+func (i invalid) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: "invalid", LID: "invalid", Name: "invalid"}
```
before this struct doesn't implement method GetName, so it's not needed to return `Name` for such cases.

IvanRN replied:
removed

artlo commented on api_test.go:
```diff
@@ -37,25 +37,27 @@ func (m *requestURLResolver) SetRequest(r http.Request) {
-func (p Post) GetID() string {
-	return p.ID
+func (p Post) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: p.ID, LID: p.LID, Name: "posts"}
```
the same

IvanRN replied:
removed

artlo commented on api_test.go:
```diff
@@ -187,30 +189,33 @@ func (p Post) GetReferencedStructs() []jsonapi.MarshalIdentifier {
-func (c Comment) GetID() string {
-	return c.ID
+func (c Comment) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: c.ID, LID: c.LID, Name: "comments"}
```
the same

IvanRN replied:
removed

artlo commented on api_test.go:
```diff
@@ -187,30 +189,33 @@ func (p Post) GetReferencedStructs() []jsonapi.MarshalIdentifier {
-func (u User) GetID() string {
-	return u.ID
+func (u User) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: u.ID, LID: u.LID, Name: "users"}
```
the same

IvanRN replied:
removed

artlo commented on api_test.go:
```diff
@@ -187,30 +189,33 @@ func (p Post) GetReferencedStructs() []jsonapi.MarshalIdentifier {
-func (b Banana) GetID() string {
-	return b.ID
+func (b Banana) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: b.ID, LID: b.LID, Name: "bananas"}
```
the same

IvanRN replied:
removed

artlo commented on examples/model/model_user.go:
```diff
@@ -18,13 +19,14 @@ type User struct {
-func (u User) GetID() string {
-	return u.ID
+func (u User) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: u.ID, LID: u.LID, Name: "users"}
```
the same

IvanRN replied:
removed

artlo commented on examples/model/model_chocolate.go:
```diff
@@ -1,19 +1,23 @@
-func (c Chocolate) GetID() string {
-	return c.ID
+func (c Chocolate) GetID() jsonapi.Identifier {
+	return jsonapi.Identifier{ID: c.ID, LID: c.LID, Name: "chocolates"}
```
the same

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -26,22 +26,35 @@ func (m MagicID) String() string {
-func (c Comment) GetID() string {
-	return fmt.Sprintf("%d", c.ID)
+func (c Comment) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", c.ID), Name: "comments"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -79,27 +93,37 @@ func (c Comment) GetReferencedStructs() []MarshalIdentifier {
-func (u User) GetID() string {
-	return fmt.Sprintf("%d", u.ID)
+func (u User) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", u.ID), Name: "users"}
```
the same

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -123,37 +153,65 @@ type ErrorIDPost struct {
-func (s ErrorIDPost) GetID() string {
-	return ""
+func (s ErrorIDPost) GetID() Identifier {
+	return Identifier{ID: "", LID: "", Name: "error-id-posts"}
```
the same

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -123,37 +153,65 @@ type ErrorIDPost struct {
-func (c Post) GetID() string {
-	return fmt.Sprintf("%d", c.ID)
+func (c Post) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", c.ID), Name: "posts"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -272,12 +353,17 @@ func (c *Post) SetReferencedStructs(references []UnmarshalIdentifier) error {
-func (p AnotherPost) GetID() string {
-	return fmt.Sprintf("%d", p.ID)
+func (p AnotherPost) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", p.ID), Name: "another-posts"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -301,22 +387,24 @@ func (p AnotherPost) GetReferencedIDs() []ReferenceID {
-func (z ZeroPost) GetID() string {
-	return z.ID
+func (z ZeroPost) GetID() Identifier {
+	return Identifier{ID: z.ID, LID: z.LID, Name: "zero-posts"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -361,86 +449,110 @@ func (q Question) GetReferencedStructs() []MarshalIdentifier {
-func (i Identity) GetID() string {
-	return fmt.Sprintf("%d", i.ID)
+func (i Identity) GetID() Identifier {
+	id := Identifier{ID: fmt.Sprintf("%d", i.ID), Name: "identities"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -361,86 +449,110 @@ func (q Question) GetReferencedStructs() []MarshalIdentifier {
-func (n *NumberPost) SetID(ID string) error {
-	n.ID = ID
+func (n NumberPost) GetID() Identifier{
+	return Identifier{ID: n.ID, LID: n.LID, Name: "number-posts"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -361,86 +449,110 @@ func (q Question) GetReferencedStructs() []MarshalIdentifier {
-func (s SQLNullPost) GetID() string {
-	return s.ID
+func (s SQLNullPost) GetID() Identifier {
+	return Identifier{ID: s.ID, LID: s.LID, Name: "sqlNullPosts"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed in other cases when GetName was not implemented initially, here (SQLNulPost) it actually was here before my changes

artlo commented on jsonapi/fixtures_test.go:
```diff
@@ -561,51 +661,45 @@ func (n CustomMetaPost) GetCustomMeta(linkURL string) map[string]Meta {
-func (i Image) GetID() string {
-	return i.ID
+func (i Image) GetID() Identifier {
+	return Identifier{ID: i.ID, LID: i.LID, Name: "images"}
```
No need to define Name, since GetName was not implemented before

IvanRN replied:
removed

artlo commented on jsonapi/marshal.go:
```diff
@@ -446,9 +457,9 @@ func marshalStruct(data MarshalIdentifier, information ServerInformation) (*Docu
-	entityName, ok := data.(EntityNamer)
-	if ok {
-		return entityName.GetName()
+	identifier, ok := data.(MarshalIdentifier)
```
I think that getStructType should accept MarshalIdentifier as argument, so you no need to cast type here

IvanRN replied:
I need to perform casting to MarshalIdentifier in getStructType or in the checkType. Should I return to double casting in the checkType?
